### ISSUE TITLE
Support for `timeout` and `eof`

### ISSUE THREAD

**Issue by ryanbmilbourne** [spectcl/spectcl#15]
Add handling of the special parameter keywords that TCL Expect supports.  We may also need to support `abort` and `close`, but I am not sure yet if those are special keywords in the TCL Expect or if they're only supported in the C code.
* [ ] timeout
* [ ] eof

**Comment by ryanbmilbourne** [comment]
In order to properly support `timeout` and `eof` as 'regular' expect cases and not something to be handled via an event listener, I'm implementing a callback that is called after the expectations have been handled.  Something like this:

``` javascript
session.expect([
    /hello/, function(match, cb){
        console.log('hello!')
        cb()
    },
    session.TIMEOUT, function(match, cb){
        cb(new Error('timed out!'))
    }
], function(err){
    if(err){
        console.log('expect error: %s',err)
    }
    console.log('all done')
})
```

The callback will be called at the end of the function that is called when an expect is matched.  It will also be called when the session stream ends (eof) and when timeout occurs.  In both cases, timeout or eof handlers will be called before the final callback is called.

This control flow is a little bit like `async`'s `parallel` or `eachSeries` control flow, where a final callback is called with an Error object once the work is done.

This also gives us the ability to handle the results of the expect statement in one place (the final callback), rather than handling in the individual handler functions:

``` javascript
session.spawn('ssh 1.2.3.4')
session.expect([
    /user/i, function(){
        session.send('foo\r')
        return session.EXP_CONTINUE
    },
    /ssword/i, function(){
        session.send('bar\r')
        return session.EXP_CONTINUE
    },
    />/, function(match, cb){
        cb()
    },
    /invalid\ password/i, function(match, cb){
        cb(new Error('invalid password!')
    },
    session.EXP_TIMEOUT, function(match, cb){
        cb(new Error('timeout'))
    }
], function(err){
    if(err){
        console.log(err)
    }
    console.log('all done')
})
```

TCL Expect's `expect` command returns when a match is found, a timeout period has passed, or if eof is seen .  The final callback would be called in each of those situations, thus matching TCL Expect's behavior.

Thoughts @gcochard?

**Comment by gcochard** [comment]
If you do another expect call within this one, will this still work? For example:

```js
session.spawn('ssh 1.2.3.4')
session.expect([
    /user/i, function(){
        session.send('foo\r')
        return session.EXP_CONTINUE
    },
    /ssword/i, function(){
        session.send('bar\r')
        return session.EXP_CONTINUE
    },
    />/, function(match, cb){
        session.send('\r')
        session.expect([
            />/, function(match, inner_cb){
                session.send('exit\r')
                inner_cb()
            }
        ], function(err){
            if(err){
                console.log(err)
                cb(err)
            }
            console.log('all done with inner')
            cb()
        })
    },
    /invalid\ password/i, function(match, cb){
        cb(new Error('invalid password!')
    },
    session.EXP_TIMEOUT, function(match, cb){
        cb(new Error('timeout'))
    }
], function(err){
    if(err){
        console.log(err)
    }
    console.log('all done')
})
```

Just wondering whether this will work or should be documented as an abuse of spectcl and specifically advised against.

**Comment by ryanbmilbourne** [comment]
I think if we want to use the final callback, the stipulation would have to be that the expect statement has not been completed until its final callback is called.  As such, the sample above would emit an `error` and should be documented and advised against.

That being said, in TCL Expect this is valid:

``` tcl
spawn node "--interactive" 
expect {
    ">" {
        exp_send "console.log(\"testing\")\r"
        expect {
            ">" {
                exp_send "process.exit()\r"
                puts "output was:\n$expect_out(buffer)"
            }
        }
    }
}
```
Technically the expect-y bit (watch input stream for a match) has been completed when we enter the inner callback...so I could see it going either way.

What do you think?  I'm leaning towards the inner-callback sample being valid, although it would reduce readability somewhat.

**Comment by ryanbmilbourne** [comment]
After reading into how `full_buffer` and `match_max` work in Tcl Expect, I believe that it's a large enough task to deserve its own issue.  It turns out that `match_max` is fuzzier than first thought, and will require additional effort/thought.  I'm removing `full_buffer` from the scope of this issue, and into its own.
